### ISSUE TITLE
run-errand view of logs isn't easy (for me) to read

### ISSUE THREAD

**Issue by drnic** [cloudfoundry/bosh-cli#153]
![screen shot 2017-02-27 at 10 53 49 am](https://cloud.githubusercontent.com/assets/108/23345506/25a25982-fcdb-11e6-9009-981426a31259.png)

![screen shot 2017-02-27 at 10 55 27 am](https://cloud.githubusercontent.com/assets/108/23345519/47f36288-fcdb-11e6-8a47-f0c99f55742c.png)

I assume that the two column view is useful for someone of the core team for the errands that they are running; but the above samples show that there is something awkward in their output for my errand. Am I doing something wrong; or do I need a huge monitor?

Perhaps interlacing stdout/stderr into a single column is useful.

**Comment by drnic** [comment]
@cppforlife ping; no pivotal tracker bot

**Comment by jhunt** [comment]
Agree with @drnic - I would much rather see stdout / stderr interleaved, with either a prefix (`STDOUT: ` / `STDER: `), or colorized, or sent to real stdout / stderr streams for redirection, or some combination thereof.

**Comment by cppforlife** [comment]
we are planning to flip the table so that it looks like...

```
exit   ...
stdout ...
stderr ...

exit   ...
stdout ...
stderr ...
```

some of the work that @monkeyherder is working on right now will make this possible.

**Comment by jhunt** [comment]
is there a reason for not interleaving stderr / stdout?

**Comment by dpb587-pivotal** [comment]
@jhunt errands pipe their logs to the filesystem (`run.stdout.log`, `run.stderr.log`) and then the files are retrieved once the errand is done, so not currently possible to interleave with `run-errand`.

**Comment by jhunt** [comment]
The separation of stdout / stderr on the filesystem for things like post-start scripts and errands has always bothered me as well.  Once separated, it is (as you also mentioned) impossible to correlate them.  Would it be possible to find another solution?

**Comment by drnic** [comment]
I'd be interested to hear from the pro-stderr-separate-from-stdout camp: what are the benefits of having these two pipes split and decoupled chronologically; verses unifying them into one pipe where they are now chronologically interlaced. Which I personally like.

**Comment by dpb587-pivotal** [comment]
I don't know the "official" stance, but personally I think they should be kept separate for the same reasons that unix split things into two streams - stdout should be for the normal output of a program, whereas stderr is intended for errors and diagnostic messages.

So, I should be writing my processes such that stdout is consumable, chainable, and predictable so others can depend on its output. Then I treat stderr as being for exceptions[/diagnostics/errors/user debugging] where my downstream consumers should not need to worry about differentiating errors from my process' output and real purpose. In general, the stderr output should be indicating what the problem is and have sufficient information to identify an issue without needing to correlate streams.

Errands are a good example where its output may be consumed by something else externally. The other lifecycle scripts aren't likely to have dependents, but I think it's easier to have one process output method defined in the ecosystem.

For processes which make an intentional effort to unify stdout/stderr message formats (like using debug/info/error levels), they can typically use a single stream safely.

**Comment by jhunt** [comment]
In UNIX, the two streams are separate but interleaved -- both print to the same terminal device.  It's only when you start to redirect them that you get split behavior (opt-in), which I wholeheartedly support.  What we have right now is the split behavior without the ability to view/review them in a combined way, and once you've split the streams you cannot go back to the combined view with any fidelity.

**Comment by jfmyers9** [comment]
We have since updated the output of `run-errand` to be formatted in a vertical table as suggested.

I am going to close this issue due to inactivity. If you would like to continue the conversation about interleaving stdout/stderr, please feel free to create a new issue.
